A system to build multiple MooseDocs configurations simultaneously

crswong888 opened the pull request:
@aeslaughter, please take a look at this when you get the chance. To demonstrate the results you can run something like the following, which builds the test site, workshop, and a new test site all at the same time with cross-linking capabilities:

```
cd ~/projects/moose/python/MooseDocs/test
./moosedocs.py build --serve --subconfigs ../../../tutorials/darcy_thermo_mech/doc/config.yml subconfig/config.yml
```

I still need to set up the live server routines and a comprehensive testing system. Please make any requests for changes that you have at this point.

(closes #18137) (refs #18136)

## Post-merge Update

The command listed above is no longer valid. To demo this new feature, run the following instead:

```
cd ~/projects/moose/python/MooseDocs/test
./moosedocs.py build --serve --config config.yml subsite_config.yml
```

Although, all files that end with 'config.yml' are now the default list of configurations, so the above is equivalent to `./moosedocs.py build --serve`.

crswong888 committed:
implemented routines for creating multiple Translator objects in a single build job by specifying "subconfigurations" and made other miscellaneous modifications to MooseDocs

(refs #18137)

crswong888 committed:
this is a MooseDocs content directory for testing the new subconfiguration capabilities

(refs #18137)

crswong888 committed:
Restructured workshop contents so that files are properly written when it is built by a subtranslator

(refs #18137)

crswong888 committed:
using new format for 'home' key

(refs #18136)

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/768905/) on f8a7e9e wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/18211/site)

This comment will be updated on new commits.

aeslaughter reviewed (changes requested):
Made some comments on the main changes, this is a great start. Thanks for the work on this, it will be really great when it is done. It will be another strong feature that will set MooseDocs apart from other similar tools.

crswong888 committed:
setting up subconfigs for workshop and test subsite and handling various conflicts that emerged

(refs #18137 #17900)

crswong888 committed:
Placing less emphasis on the "primary" translator - all translators should share the exact same content pool and everything should build the same regardless of which order the configs come in

(refs #18137)

crswong888 committed:
This commit implements the following:
    - More methods to make all generated translator objects be treated equally, there's no such thing as a "primary" translator.
    - Made it so that executioners can be called for different build steps so that all translators tokenize before rendering
    - Nominally refactored various parts of MooseDocs

(refs #18137)

crswong888 commented on python/MooseDocs/commands/build.py:
```diff
@@ -136,7 +139,7 @@ def _getPage(self, source):
                 if isinstance(page, pages.Source):
                     page.output_extension = self._translator.renderer.EXTENSION
-                self._translator.addContent(page)
+                self._translator.addContent(page) # shoudldn't this be addPage()???
```
I'll double-check this once I get into modifying the live server

crswong888 commented on python/MooseDocs/base/executioners.py:
```diff
@@ -123,11 +122,19 @@ def init(self, destination):
+    def addPage(self, page, set_uid):
         """Add a Page object to be Translated."""
-        page._Page__unique_id = len(self._page_objects)
+        if set_uid:
+            page._Page__unique_id = len(self._page_objects)
```
we shall handle this one at a later time

crswong888 committed:
addressing some PR comments and implementing the read, tokenize, render, and write options in all executioners

(refs #18137)

milljm commented:
@crswong888  I am looking into the conda issues. It has to do with the recent use of `mamba`

crswong888 commented:
@milljm No worries! Do you mind if I do another force-push and restart these tests though?

crswong888 commented:
@aeslaughter, this is ready for your review. The top 4 commits are new.

Basically, there's no such thing as a "primary" translator anymore, each are treated equally and the result should be the same no matter what order they are loaded in. Each configuration should specify unique content, otherwise, `load_configs()` will raise an exception, and every translator shares the same exact pool of content. Pretty much everything new I did is in support of this idea that no translator should be special. Additonally, I addressed your comments in the best way that I could.

crswong888 committed:
This commit implements the following MooseDocs improvements:
    - The MooseDocs watcher class was modified to handle multiple translators
    - Executioners no longer modify Page.uid to allow for indicial access to data containers, but rather use the UIDs directly as a dictionary key
    - New pages may now be safely added to translators during any part of the build process

(refs #18137)

crswong888 committed:
Reverted back from storing css/js filenames in the global attributes, since this turned out to be extremely slow and required a multiprocessing lock to prevent race conditions. Instead, I am storing common files in a renderer attribute and per-page files in the page attributes, for which the latter are properly communicated across processors. I am not using a Storage() object as it was before, as these don't pickle well in parallel.

(refs #18137)

crswong888 committed:
Enabling tests for Queue and Pipe executioners. These are now stable as part of work done for #18137

(refs #12528 #15887)

crswong888 commented:
The live server is now operational.

Additionally, executioners store translator data in dictionaries now and do not modify the Page.uid attribute. Pages may be added during any part of the build procedure and are no longer blocked, including during a live serve. Although, pages added during a live serve of multiple configurations are ignored since it is not possible to determine which translator should be responsible.

It seems that all executioners are now stable and so I have enabled their tests. Whether or not this is because of changes made here is not clear. I am also not sure how to reproduce the steps described in #15887.

crswong888 committed:
fixed conflicts with work done for #18137 by adding content directly from large media

(refs #18473)

milljm commented:
@crswong888 sorry, it seems I ignored one of your comments! Wasn't intentional.

I am here because this is all very interesting. I am thinking about the ways we can take advantage of this in Civet...

crswong888 commented:
@milljm No worries. Let me know what your ideas are and I can help.

The possibilities for our web design have grown by a lot with this new system. A single page could have its own unique configuration entirely independent from others. This is by no means limited to just slides. For example, take a look at `python/MooseDocs/test/subsite_config.yml`
